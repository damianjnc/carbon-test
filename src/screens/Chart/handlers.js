// @flow

type SetGradientColor = (canvas: any, color: string) => void

export const handleSetGradientColor = (canvas: any, color: string) => {
  const ctx = canvas.getContext('2d')
  const gradient = ctx.createLinearGradient(0, 0, 600, 550)

  gradient.addColorStop(0, color)
  gradient.addColorStop(0.95, 'rgba(133, 122, 144, 0.5)')

  return gradient
}

export const handleGetChartData = (data: any) => (
  setGradientColor: SetGradientColor
) => (canvas: any) => {
  if (data.datasets) {
    let colors = ['rgba(255, 0, 255, 0.75)', 'rgba(0, 255, 0, 0.75)']
    data.datasets.forEach((set, index) => {
      set.backgroundColor = setGradientColor(canvas, colors[index])
      set.borderColor = 'white'
      set.borderWidth = 2
    })
  }

  return data
}
