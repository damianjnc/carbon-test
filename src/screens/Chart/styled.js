// @flow

import styled from 'styled-components'

export const Wrapper = styled.div`
  width: 80%;
  display: flex;
  flex-direction: column;
  align-items: center;
`