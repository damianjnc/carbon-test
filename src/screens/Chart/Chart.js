// @flow

import React from 'react'
import { Line } from 'react-chartjs-2'

import { Wrapper } from './styled'
import { type ComponentProps } from '.'

export const Chart = ({ data }: ComponentProps) => (
  <Wrapper>
    <Line
      options={{
        responsive: true,
        title: {
          display: true,
          text: 'Top Polluters 🌍',
          fontSize: 55
        }
      }}
      data={data}
    />
  </Wrapper>
)

export default Chart
