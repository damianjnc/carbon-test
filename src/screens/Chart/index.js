// @flow

import { useEffect, useState } from 'react'

import prepareComponent from 'hocs/prepareComponent'
import Chart from './Chart'
import {
  handleGetChartData,
  handleSetGradientColor
} from 'screens/Chart/handlers'
import JsonData from 'screens/Chart/data/TopPolluters.json'

export type ComponentProps = {
  data: () => void
}

export const usePrepareComponent = (): ComponentProps => {
  const [data, setData] = useState({})

  useEffect(() => {
    const json = JSON.stringify(JsonData)
    const object = JSON.parse(json)

    const names = object.map(item => item.Name)
    const globalEmmissionsPast = object.map(item => {
      const index = item['Global emissions 1965-2017']
        .split('')
        .findIndex(element => element === 'm')

      const value = item['Global emissions 1965-2017'].slice(0, index)

      const removeCommas = Number(value.replace(/,/g, ''))

      return removeCommas
    })

    const globalEmmissionsFuture = object.map(item => {
      let value

      if (item['Projected future emissions 2018-30']) {
        const index = item['Projected future emissions 2018-30']
          .split('')
          .findIndex(element => element === 'm')

        value = item['Projected future emissions 2018-30'].slice(0, index)
      } else if (item['Projected emissions 2018-30']) {
        const index = item['Projected emissions 2018-30']
          .split('')
          .findIndex(element => element === 'm')

        value = item['Projected emissions 2018-30'].slice(0, index)
      } else {
        value = '0'
      }

      const removeCommas = Number(value.replace(/,/g, ''))

      return removeCommas
    })

    setData({
      labels: names,
      datasets: [
        {
          label: 'Global emissions 1965-2017',
          backgroundColor: 'rgba(255, 0, 255, 0.75)',
          data: globalEmmissionsPast
        },
        {
          label: 'Projected emissions 2018-30',
          backgroundColor: 'rgba(0, 255, 0, 0.75)',
          data: globalEmmissionsFuture
        }
      ]
    })
  }, [])

  return {
    data: handleGetChartData(data)(handleSetGradientColor)
  }
}

export default prepareComponent<*, ComponentProps>(usePrepareComponent)(Chart)
