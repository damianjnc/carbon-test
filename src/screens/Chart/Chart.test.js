// @flow

import React from 'react'
import { shallow } from 'enzyme'

import { Chart } from './Chart'

jest.mock('react-chartjs-2', () => ({ Line: 'Line' }))
jest.mock('./styled', () => ({ Wrapper: 'Wrapper' }))
jest.mock('.', () => ({}))

describe('screens > Chart', () => {
  describe('Chart', () => {
    it('renders', () => {
      const props = {
        data: jest.fn()
      }

      const wrapper = shallow(<Chart {...props} />)

      expect(wrapper).toMatchSnapshot()
    })
  })
})
