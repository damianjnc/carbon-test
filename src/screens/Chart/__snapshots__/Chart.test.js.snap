// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screens > Chart Chart renders 1`] = `
<Wrapper>
  <Line
    data={[MockFunction]}
    options={
      Object {
        "responsive": true,
        "title": Object {
          "display": true,
          "fontSize": 55,
          "text": "Top Polluters 🌍",
        },
      }
    }
  />
</Wrapper>
`;
